[monitoring] Allow to store IP addresses of interfaces in DeviceData.data #30

nepython opened the pull request:
~~This is still WIP. The following need to be done:-~~
* [x] show the addresses in the device status tab: https://github.com/openwisp/openwisp-monitoring/blob/master/openwisp_monitoring/device/templates/admin/config/device/change_form.html
* [x] add tests to ensure bogus addresses will be rejected
* [x] Creating a PR in `monitoring-lua` as mentioned in the issue

**PS**: The JSON type for `family` was provided as `integer` [here](http://netjson.org/rfc.html#addresses1) but it appeared visibly as `string` to me

Closes #30

nemesisdesign commented on openwisp_monitoring/device/models.py:
```diff
@@ -74,6 +74,11 @@ def data_user_friendly(self):
+            if 'addresses' in interface:
+                data['addresses'] = {}
+                for count, address in enumerate(interface['addresses']):
+                    data['addresses']['address' + str(count)] = address['address']
```
unless I'm missing something, I would keep this data structure as is and just display it as a table in the device status tab.

nepython commented:
@nemesisdesign, I tried out few styles for the table (I think, I am pretty bad at CSS :sweat_smile: but I am improving). Can you please tell me is this what you have in mind.
![DeviceData_table](https://user-images.githubusercontent.com/54471024/81039316-385ccb00-8e98-11ea-918a-f7d362594c90.png)
I thought it would be appropriate to highlight the ones with dhcp client.

nemesisdesign commented on openwisp_monitoring/device/static/monitoring/css/monitoring.css:
```diff
@@ -47,3 +47,25 @@ td.field-health_status,
 .health-problem{ color: orange }
 .health-critical{ color: red }
+
+table {
```
it's not a good idea to use `table` without referring it to a class or id, because it will be applied to any table, although this css is loaded only in a few pages so it won't affect the rest of the application, but do not get used to bad habits.

nepython replied:
I was going to change it as I noticed with it the device change list displays terribly :laughing:, expected after all it is also a `table`

nemesisdesign commented on openwisp_monitoring/device/templates/admin/config/device/change_form.html:
```diff
@@ -120,6 +120,29 @@ <h2>{% trans 'Interface status' %}: {{ interface.name }}</h2>
+                <h2>{% trans 'ADDRESSES' %}:</h2>
+                  <table style="width:100%">
+                    <tr>
+                      <th>Address</th>
```
Should be `{% translate 'Address' %}`, same applies to the other strings.

nepython commented:
@nemesisdesign, the PR is not yet ready. I just read your `gitter` advice to use https://github.com/openwisp/openwisp-config/blob/master/openwisp-config/files/lib/openwisp/net.lua inorder to retrieve the addresses, missed to read it somehow. I tested it by introducing the data via django shell :sweat_smile:. Let me try this and update the PR accordingly

PabloCastellano commented on openwisp_monitoring/device/schema.py:
```diff
@@ -142,6 +142,23 @@
+                            "properties": {
+                                "proto": {"type": "string"},
+                                "family": {"type": "string"},
+                                "address": {"type": "string", "format": "email"},
```
Is it a mistake?

```
                    "ip_address": {
                        "type": "string",
                        "anyOf": [{"format": "ipv4"}, {"format": "ipv6"}],
                    },
```

should work better

nepython replied:
Well, I checked up all formats and nothing was validating so I was figuring it out then. Now it validates, just read the jsonschema docs again and found out `format_checker` isn't enabled by default, changed it and it validates all formats perfectly. 
Thanks, for the suggestion. Any reason why you prefer `anyOf` over `oneOf`, I haven't used **jsonschema** much, hence curious :smiley:

PabloCastellano replied:
For this case the result is the same so it doesn't matter. Probably anyOf is faster because of short-circuit evaluation.

nepython commented:
@nemesisdesign, it is ready for review. I am working on getting the addresses from the interface currently, it would be great if you can pass few instructions to get started with [this](https://github.com/openwisp/openwisp-config/blob/master/openwisp-config/files/lib/openwisp/net.lua) and I have updated the styling slightly as in the screenshot :point_up:.

nepython commented:
@nemesisdesign, rebased the PR and tested it by using the script **netjson-monitoring-wan.lua** as config, it works perfectly :heart_eyes:. I will open a PR soon in `monitoring-lua` as mentioned in the issue #30

PabloCastellano commented on openwisp_monitoring/device/tests/test_models.py:
```diff
@@ -223,6 +238,19 @@ def test_device_deleted(self):
         else:
             self.fail('metric was not deleted')
 
+    def test_bad_address_fail(self):
```
Can you add another test for ipv6?

nepython replied:
@PabloCastellano, is it required? In this test, I have put the `ip-address` as some bogus string and the output message mentions `ipv4` specifically as it is the first of the `format` in the two possibilities (I am unsure though, I might be wrong). Do you mean the tests should also validate that the `ip-address` have the format as `family`. If yes, I don't think it can be done with **json-schema** so I will have to try something else.

PabloCastellano replied:
I think we don't need it anymore :+1:

PabloCastellano commented on openwisp_monitoring/device/schema.py:
```diff
@@ -145,6 +145,26 @@
+                            "title": "Address",
+                            "type": "object",
+                            "properties": {
+                                "proto": {"type": "string"},
```
Possible values for `proto` so far are: `static`, `dhcp`, `dhcpv6`. @nemesisdesign do you think we should validate that proto contains any of these values?

nemesisdesign replied:
netjson has only `static` and `dhcp`, for understanding if it's ipv4 or ipv6 we use `family`. Yes I think it's useful to enforce this as well, we can use `enum` or anything similar

PabloCastellano replied:
So we need to send dhcp from the lua script even if ubus returns dhcpv6, right?

nepython replied:
Same query :smile:

nemesisdesign replied:
What do you mean with ubus returns dhcpv6 exactly?

I think the answer to the question is yes, proto is `dhcp` or `static`, family is `ipv4` or `ipv6`.

PabloCastellano replied:
@nemesisdesign I meant that `ubus call network.interface dump` returns `dhcp` and `dhcpv6` as possible values in the `proto` field so we cannot just send the value from ubus

nepython commented:
@nemesisdesign, I had rebased previously when #63 got merged. Updated the style, now it is same as `arp-table` besides two minor changes, I kept a `margin-top` for the table as there wasn't any space between `form-row` and the `table` and highlighted the ones having `dhcp` client :blush:

nemesisdesign commented on openwisp_monitoring/device/templates/admin/config/device/change_form.html:
```diff
@@ -141,6 +141,31 @@ <h2>{% trans 'Interface status' %}: {{ interface.name }}</h2>
               </div>
+              <div class = "device_addresses">
+                {% if interface.addresses %}
+                  <h2>{% trans 'ADDRESSES' %}:</h2>
```
there's no need to write it uppercase, that's CSS's job.

The fact that the table contain IP addresses is self explanatory so this heading is not needed. You can remove it.

nemesisdesign commented on openwisp_monitoring/device/templates/admin/config/device/change_form.html:
```diff
@@ -141,6 +141,31 @@ <h2>{% trans 'Interface status' %}: {{ interface.name }}</h2>
                   {% endif %}
                   {% endif %}
               </div>
+              <div class = "device_addresses">
```
the css class should use the dash for consistency.
The div is not needed here, you can elminate the div and move the class to the table element

nemesisdesign commented on openwisp_monitoring/device/templates/admin/config/device/change_form.html:
```diff
@@ -141,6 +141,31 @@ <h2>{% trans 'Interface status' %}: {{ interface.name }}</h2>
+                  <h2>{% trans 'ADDRESSES' %}:</h2>
+                    <table>
+                      <thead>
+                        <tr>
```
be consistent with indentation, 4 space indentation is being used in this file, so follow it please. Adopt the same methodology on any file you work on please.
